runtime/compiler/optimizer/J9EstimateCodeSize.cpp

```diff
@@ -1472,6 +1486,15 @@ TR_J9EstimateCodeSize::realEstimateCodeSize(TR_CallTarget *calltarget, TR_CallSt
       bci.prepareToFindAndCreateCallsites(blocks, flags, callSites, &cfg, &newBCInfo, _recursionDepth, &callStack);
       bool iteratorWithState = (inlineArchetypeSpecimen && !mhInlineWithPeeking) || inlineLambdaFormGeneratedMethod;
 
+// in JDK21+, iterating bytecodes of MemorySegment methods with state is necesssary to obtain the segment view
+// VarHandle object through folding ValueLayouts$AbstractValueLayout.accessHandle() if the VarHandle has been created
+#if JAVA_SPEC_VERSION >= 21
+      if (callerName && !strncmp(callerName, "java/lang/foreign/MemorySegment", 31))
```

Comment:
Currently `callerName` is non-null only when `traceInlining` or `debugInliner` is enabled, so I think you'll find that this isn't usually kicking in. It would look like it's working because when you turn on tracing to check, the tracing makes it work